upbc/protoc-gen-upb.cc

```diff
@@ -418,294 +418,362 @@ void GenerateExtensionInHeader(const protobuf::FieldDescriptor* ext,
+  } else {
+    output(
+        R"cc(
+          UPB_INLINE $0 $1_$2(const $1* msg) {
+            return *UPB_PTR_AT(msg, $3, $0);
           }
         )cc",
-        MessageName(message), MessageInit(message));
+        CTypeConst(field), msg_name, field->name(),
+        GetSizeInit(layout.GetFieldOffset(field)));
   }
+}
 
-  for (int i = 0; i < message->real_oneof_decl_count(); i++) {
-    const protobuf::OneofDescriptor* oneof = message->oneof_decl(i);
-    std::string fullname = ToCIdent(oneof->full_name());
-    output("typedef enum {\n");
-    for (int j = 0; j < oneof->field_count(); j++) {
-      const protobuf::FieldDescriptor* field = oneof->field(j);
-      output("  $0_$1 = $2,\n", fullname, field->name(), field->number());
-    }
+void GenerateGetters(const protobuf::FieldDescriptor* field,
+                     const MessageLayout& layout, absl::string_view msg_name,
+                     Output& output) {
+  if (field->is_map()) {
+    GenerateMapGetters(field, layout, msg_name, output);
+  } else if (field->containing_type()->options().map_entry()) {
+    GenerateMapEntryGetters(field, msg_name, output);
+  } else if (field->is_repeated()) {
+    GenerateRepeatedGetters(field, layout, msg_name, output);
+  } else if (field->real_containing_oneof()) {
+    GenerateOneofGetters(field, layout, msg_name, output);
+  } else {
+    GenerateScalarGetters(field, layout, msg_name, output);
+  }
+}
+
+void GenerateMapSetters(const protobuf::FieldDescriptor* field,
+                        const MessageLayout& layout, absl::string_view msg_name,
+                        Output& output) {
+  const protobuf::Descriptor* entry = field->message_type();
+  const protobuf::FieldDescriptor* key = entry->FindFieldByNumber(1);
+  const protobuf::FieldDescriptor* val = entry->FindFieldByNumber(2);
+  output(
+      "UPB_INLINE void $0_$1_clear($0 *msg) { _upb_msg_map_clear(msg, "
+      "$2); "
+      "}\n",
+      msg_name, field->name(), GetSizeInit(layout.GetFieldOffset(field)));
+  output(
+      "UPB_INLINE bool $0_$1_set($0 *msg, $2 key, $3 val, upb_Arena *a) "
+      "{ "
+      "return _upb_msg_map_set(msg, $4, &key, $5, &val, $6, a); }\n",
+      msg_name, field->name(), CType(key), CType(val),
+      GetSizeInit(layout.GetFieldOffset(field)),
+      key->cpp_type() == protobuf::FieldDescriptor::CPPTYPE_STRING
+          ? "0"
+          : "sizeof(key)",
+      val->cpp_type() == protobuf::FieldDescriptor::CPPTYPE_STRING
+          ? "0"
+          : "sizeof(val)");
+  output(
+      "UPB_INLINE bool $0_$1_delete($0 *msg, $2 key) { "
+      "return _upb_msg_map_delete(msg, $3, &key, $4); }\n",
+      msg_name, field->name(), CType(key),
+      GetSizeInit(layout.GetFieldOffset(field)),
+      key->cpp_type() == protobuf::FieldDescriptor::CPPTYPE_STRING
+          ? "0"
+          : "sizeof(key)");
+  output(
+      "UPB_INLINE $0 $1_$2_nextmutable($1 *msg, size_t* iter) { "
```

Comment:
maybe in some other CL we should consider hitting these with more consistent casing...